src/transforms/ampDisableCacheTransform.js

```diff
@@ -14,25 +14,89 @@
+const fetchRuntime = require('@ampproject/toolbox-runtime-fetch');
+const fs = require('fs');
+const path = require('path');
+const {hasAttribute, firstChildByTag} = require('@ampproject/toolbox-optimizer/lib/NodeUtils');
 
 /**
  * Removes the `amp` attribute from the `html` tag to disable serving
  * the page from the AMP Cache. We do this in a separate transformer
- * to still support AMP validation.
+ * to still support AMP validation. This will also download and self-host
+ * the complete AMP runtime.
+ *
+ * Note: amp-geo won't work in this mode as it's an AMP Cache feature.
+ * See https://github.com/ampproject/amphtml/blob/master/spec/amp-framework-hosting.md#amp-geo-fallback-api
  */
-const ampDisableCacheTransform = (eleventyConfig, options = {}) => {
+const ampDisableCacheTransform = async (eleventyConfig, options = {}) => {
   if (options.ampCache !== false) {
     return;
   }
+
+  /**
+   * A custom transformer that detects the used AMP runtime version and triggers the AMP runtime download.
+   */
+  class DownloadRuntime {
+    constructor(config) {
+      this.log = config.log;
+    }
+    async transform(root, params) {
+      const outputDir = options.ampRuntimeDir || '_site';
+      const html = firstChildByTag(root, 'html');
+      const head = firstChildByTag(html, 'head');
+      for (const node of head.children) {
+        if (node.tagName !== 'style') {
+          continue;
+        }
+        if (!hasAttribute(node, 'amp-runtime')) {
+          continue;
+        }
+        const ampRuntimeVersion = node.attribs['i-amphtml-version'];
+        if (!ampRuntimeVersion) {
+          continue;
+        }
+
+        const ampUrlPrefix = path.join('rtv', ampRuntimeVersion);
+        const downloadSuccess = await this.downloadRuntime(
+          ampRuntimeVersion,
+          outputDir,
+          ampUrlPrefix
+        );
+        if (!downloadSuccess) {
+          this.log.warn('Failed downloading AMP runtime.');
+          return;
+        }
+        // Runtime has been successfully downloaded, update params
+        // to rewrite the script import URLs in subsequent transformations
+        params.ampRuntimeVersion = ampRuntimeVersion;
+        params.ampUrlPrefix = ampUrlPrefix;
+      }
+    }
+
+    async downloadRuntime(ampRuntimeVersion, ouputDir, runtimeDir) {
+      const targetDir = path.join(ouputDir, runtimeDir);
+      if (fs.existsSync(targetDir)) {
+        return true;
+      }
+      this.log.info('Downloading AMP runtime version', ampRuntimeVersion);
+      return fetchRuntime.getRuntime({
+        dest: ouputDir,
+      }).status;
+    }
+  }
+
   const ampOptimizer = AmpOptimizer.create({
-    transformations: ['RemoveAmpAttribute'],
+    transformations: [DownloadRuntime, 'RemoveAmpAttribute', 'RewriteAmpUrls'],
   });
   eleventyConfig.addTransform('amp-disable-cache', async (content, outputPath) => {
-    if (!outputPath.endsWith('.html')) {
+    if (!options.isAmp(outputPath)) {
```

Comment:
Are you confident this is always defined? Why are we setting a [default param](https://github.com/ampproject/eleventy-plugin-amp/pull/32/files?short_path=915e09f#diff-0770dc4fd3493744c50d191bc447b2fbR31)?

Reply:
Did you implement this in AmpConfig?

Reply:
I think the default param is legacy and we should remove it.

Reply:
Removed the default param. Good catch!